tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Tensor.java

```diff
@@ -26,6 +26,16 @@
 import java.nio.LongBuffer;
 import java.util.Arrays;
 import java.util.HashMap;
+import org.tensorflow.internal.c_api.TF_Tensor;
+import org.tensorflow.util.Shape;
+import org.tensorflow.types.TBool;
+import org.tensorflow.types.TDouble;
+import org.tensorflow.types.TFloat;
+import org.tensorflow.types.TInt32;
+import org.tensorflow.types.TInt64;
+import org.tensorflow.types.TString;
+import org.tensorflow.types.TUInt8;
+import org.tensorflow.types.family.TType;
 
 /**
  * A statically typed multi-dimensional array whose elements are of a type described by T.
```

Comment:
Can we more strongly type `Tensor` to `Tensor<T extends TType>`?

Reply:
That is what I tried at first but to discover than this enforcement became quickly viral and that it needs to be everywhere in our methods, classes, wrappers, frameworks and, ultimately,  user code. 

But since `TType` does not expose any method and since we enforce it at the beginning of the chain (i.e. at the creation of the `Tensor`), I realized that there is no real gain carrying it afterward. We just want to make sure that tensors are created using one of our supported data types.

I think our users will appreciate that we relax a bit this rule.

Reply:
The advantage of doing it is that the erased type bound of `Tensor` becomes `Tensor<TType>` which makes it tricker for people to skirt the generics to do things that they shouldn't be doing. I agree it seems like it's everywhere in the codebase, and it does ripple a little into user code, but there is some benefit to it.

Reply:
I'm aware of the benefits of enforcing these safety checks, I've always been a defender of it, but I think we can make an exception here. I don't care that much of the verbosity in our code base but I'm concerned about the user code. 

I guess writing down a quick example may help us to take the right decision. Meanwhile, I'm still not sure to understand what wrong can happen if we leave the generic type parameter unbounded after creating the tensor, other than not being a good practice. Do you have a specific scenario in mind?